src/components/cuda/cupti_profiler.c

```diff
@@ -0,0 +1,1564 @@
+
+    gpu_ctl->counterDataScratchBuffer.size = scratchBufferSizeParams.counterDataScratchBufferSize;
+    gpu_ctl->counterDataScratchBuffer.data = (uint8_t *) papi_calloc(gpu_ctl->counterDataScratchBuffer.size, sizeof(uint8_t));
+    if (gpu_ctl->counterDataScratchBuffer.data == NULL) {
+        ERRDBG("calloc gpu_ctl->counterDataScratchBuffer.data failed!\n");
+        return PAPI_ENOMEM;
+    }
+
+    CUpti_Profiler_CounterDataImage_InitializeScratchBuffer_Params initScratchBufferParams = {
+        .structSize = CUpti_Profiler_CounterDataImage_InitializeScratchBuffer_Params_STRUCT_SIZE,
+        .pPriv = NULL,
+        .counterDataImageSize = calculateSizeParams.counterDataImageSize,
+        .pCounterDataImage = initializeParams.pCounterDataImage,
+        .counterDataScratchBufferSize = gpu_ctl->counterDataScratchBuffer.size,
+        .pCounterDataScratchBuffer = gpu_ctl->counterDataScratchBuffer.data,
+    };
+    CUPTI_CALL(cuptiProfilerCounterDataImageInitializeScratchBufferPtr(&initScratchBufferParams), goto fn_fail;);
+
+    return PAPI_OK;
+fn_fail:
+    return PAPI_EMISC;
+}
+
+static int reset_cupti_prof_config_images(cuptip_gpu_control_t *gpu_ctl)
+{
+    COMPDBG("Entering.\n");
+    papi_free(gpu_ctl->counterDataImagePrefix.data);
+    papi_free(gpu_ctl->configImage.data);
+    papi_free(gpu_ctl->counterDataImage.data);
+    papi_free(gpu_ctl->counterDataScratchBuffer.data);
+    papi_free(gpu_ctl->counterAvailabilityImage.data);
+    return PAPI_OK;
+}
+
+static int begin_profiling(cuptip_gpu_control_t *gpu_ctl)
+{
+    COMPDBG("Entering.\n");
+    byte_array_t *configImage = &(gpu_ctl->configImage);
+    byte_array_t *counterDataScratchBuffer = &(gpu_ctl->counterDataScratchBuffer);
+    byte_array_t *counterDataImage = &(gpu_ctl->counterDataImage);
+
+    CUpti_Profiler_BeginSession_Params beginSessionParams = {
+        .structSize = CUpti_Profiler_BeginSession_Params_STRUCT_SIZE,
+        .pPriv = NULL,
+        .ctx = NULL,
+        .counterDataImageSize = counterDataImage->size,
+        .pCounterDataImage = counterDataImage->data,
+        .counterDataScratchBufferSize = counterDataScratchBuffer->size,
+        .pCounterDataScratchBuffer = counterDataScratchBuffer->data,
+        .range = CUPTI_UserRange,
+        .replayMode = CUPTI_UserReplay,
+        .maxRangesPerPass = 1,
+        .maxLaunchesPerPass = 1,
+    };
+    CUPTI_CALL(cuptiProfilerBeginSessionPtr(&beginSessionParams), goto fn_fail;);
+
+    CUpti_Profiler_SetConfig_Params setConfigParams = {
+        .structSize = CUpti_Profiler_SetConfig_Params_STRUCT_SIZE,
+        .pPriv = NULL,
+        .ctx = NULL,
+        .pConfig = configImage->data,
+        .configSize = configImage->size,
+        .minNestingLevel = 1,
+        .numNestingLevels = 1,
+        .passIndex = 0,
+        .targetNestingLevel = 1,
+    };
+    CUPTI_CALL(cuptiProfilerSetConfigPtr(&setConfigParams), goto fn_fail;);
+
+    /* User takes the resposiblity of replaying the kernel launches */
```

Comment:
Be consistent with comments, you are mixing C99 (and onwards) and C89 comments. PAPI is officially C89 compliant. I don't really care about which one you use, but be consistent with the one you pick.